dandi/dandiapi.py

```diff
@@ -321,10 +321,13 @@ def get_download_file_iter(
             f"/dandisets/{dandiset_id}/versions/{version}/assets/{uuid}/download/"
         )
 
-        def downloader():
+        def downloader(start_at=0):
             lgr.debug("Starting download from %s", url)
+            headers = None
+            if start_at > 0:
+                headers = {"Range": f"bytes={start_at}-"}
```

Comment:
I tried testing download resumption for #189, and I was unable to get it to work.  I don't think there's an easy way to adapt that code to this PR that would make it suddenly start working.

Reply:
even for `dandiapi`? it should work ...

<details>
<summary>Here is an example with `wget --continue` on the url from girder -- seems to work</summary> 

```shell
wget --continue -S --verbose https://girder.dandiarchive.org/api/v1/item/5e73fad0368d3c79a8006c13/download
--2020-09-28 14:46:00--  https://girder.dandiarchive.org/api/v1/item/5e73fad0368d3c79a8006c13/download
Resolving girder.dandiarchive.org (girder.dandiarchive.org)... 3.19.164.171
Connecting to girder.dandiarchive.org (girder.dandiarchive.org)|3.19.164.171|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 303 See Other
  Server: nginx/1.14.0 (Ubuntu)
  Date: Mon, 28 Sep 2020 18:46:29 GMT
  Content-Type: text/html;charset=utf-8
  Content-Length: 1640
  Connection: keep-alive
  Allow: DELETE, GET, HEAD, OPTIONS, PATCH, POST, PUT
  Girder-Request-Uid: a20e2811-42c5-46f2-86af-731cc92edaa9
  Location: https://dandiarchive.s3.amazonaws.com/girder-assetstore/f7/65/f765fb69bd7343798fd994103c1ff2eb?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIA3GIMZPVVACWR4QDP%2F20200928%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20200928T184629Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Security-Token=FwoGZXIvYXdzEBwaDFhTG5iIAuIRe%2Fp21yK%2FAW79vdBfa6OYxwOtF8%2FZp2VgcEo1JKJCmCf37E0AKu9riF86bpqmGM%2FQIqt89ntLzCgowprsE3LvVN7U4K%2F%2FJyCkOki7mBtEyMW7e9OnvSTS0u1KR0Pcml%2BdZeCDSUQHlOHyJOrEoaNoYjkbqEB8zrNbqhv5armELJ1ZLeAmUdg0cFB0uXbA%2BiXawhLKfo8Ps%2B2NFy1ALO%2FY397OZPW0pTYBx8yjeauxYfkBuS9jSUlu602yTo851zuveNTn%2FHypKIXeyPsFMi2dGUzktZIE0qNPff49hWxIu4ZA%2FDeda2zFEEbIPpkL0M7IObu66lnYCZylD4I%3D&X-Amz-Signature=14f1482fe7187bde16ffcba4ff92d5a6315501c2dcf3f55b07b66f6a11534f8f
  Strict-Transport-Security: max-age=63072000
Location: https://dandiarchive.s3.amazonaws.com/girder-assetstore/f7/65/f765fb69bd7343798fd994103c1ff2eb?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIA3GIMZPVVACWR4QDP%2F20200928%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20200928T184629Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Security-Token=FwoGZXIvYXdzEBwaDFhTG5iIAuIRe%2Fp21yK%2FAW79vdBfa6OYxwOtF8%2FZp2VgcEo1JKJCmCf37E0AKu9riF86bpqmGM%2FQIqt89ntLzCgowprsE3LvVN7U4K%2F%2FJyCkOki7mBtEyMW7e9OnvSTS0u1KR0Pcml%2BdZeCDSUQHlOHyJOrEoaNoYjkbqEB8zrNbqhv5armELJ1ZLeAmUdg0cFB0uXbA%2BiXawhLKfo8Ps%2B2NFy1ALO%2FY397OZPW0pTYBx8yjeauxYfkBuS9jSUlu602yTo851zuveNTn%2FHypKIXeyPsFMi2dGUzktZIE0qNPff49hWxIu4ZA%2FDeda2zFEEbIPpkL0M7IObu66lnYCZylD4I%3D&X-Amz-Signature=14f1482fe7187bde16ffcba4ff92d5a6315501c2dcf3f55b07b66f6a11534f8f [following]
--2020-09-28 14:46:00--  https://dandiarchive.s3.amazonaws.com/girder-assetstore/f7/65/f765fb69bd7343798fd994103c1ff2eb?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIA3GIMZPVVACWR4QDP%2F20200928%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20200928T184629Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Security-Token=FwoGZXIvYXdzEBwaDFhTG5iIAuIRe%2Fp21yK%2FAW79vdBfa6OYxwOtF8%2FZp2VgcEo1JKJCmCf37E0AKu9riF86bpqmGM%2FQIqt89ntLzCgowprsE3LvVN7U4K%2F%2FJyCkOki7mBtEyMW7e9OnvSTS0u1KR0Pcml%2BdZeCDSUQHlOHyJOrEoaNoYjkbqEB8zrNbqhv5armELJ1ZLeAmUdg0cFB0uXbA%2BiXawhLKfo8Ps%2B2NFy1ALO%2FY397OZPW0pTYBx8yjeauxYfkBuS9jSUlu602yTo851zuveNTn%2FHypKIXeyPsFMi2dGUzktZIE0qNPff49hWxIu4ZA%2FDeda2zFEEbIPpkL0M7IObu66lnYCZylD4I%3D&X-Amz-Signature=14f1482fe7187bde16ffcba4ff92d5a6315501c2dcf3f55b07b66f6a11534f8f
Resolving dandiarchive.s3.amazonaws.com (dandiarchive.s3.amazonaws.com)... 52.219.100.92
Connecting to dandiarchive.s3.amazonaws.com (dandiarchive.s3.amazonaws.com)|52.219.100.92|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 206 Partial Content
  x-amz-id-2: 97WjpUfcAUv3U0vYREiEf01n039ggF2GllEEaaPN7HglkuenbmBMaGBfe+po55BPRUG+7OzMUhc=
  x-amz-request-id: DH7SFM6R1P1Y9G5W
  Date: Mon, 28 Sep 2020 18:46:01 GMT
  Last-Modified: Thu, 19 Mar 2020 23:05:57 GMT
  ETag: "87e9d6f8df1e6ae5cf1a8a97089890d5-193"
  Content-Disposition: attachment; filename="sub-YutaMouse20_ses-YutaMouse20-140321_behavior+ecephys.nwb"
  x-amz-meta-uploader-ip: 18.13.53.55
  x-amz-meta-uploader-id: 5d978dc2cc10d1bc31040bcc
  x-amz-version-id: V5BK7tKKBlGCdAV75e_Q98xRAWvq5NCG
  Accept-Ranges: bytes
  Content-Range: bytes 116000638-12887868733/12887868734
  Content-Type: application/octet-stream
  Content-Length: 12771868096
  Server: AmazonS3
Length: 12887868734 (12G), 12771868096 (12G) remaining [application/octet-stream]
Saving to: ‘download’

download                       0%[                                            ] 116.83M  4.13MB/s               ^C

```
</details>

As for how to test -- we could download a full file (from 000027 test dataset), rename back into `.dandidownload` one with necessary metadata and test different scenarios:

1. strip all of it leaving 0 sized
2. strip some (random?) % simulating incomplete download
3. do not strip anything - so downloader would just need to rename into the target

Reply:
When renaming the file back into `.dandidownload`, the file's checksum will also need to be stored in the directory.  How should that be fetched or determined?  If I should just digest the file's contents, what algorithm should be used?  Alternatively, since the test always uses 000027, should the checksum be hardcoded into the test, and, if so, what is its value?

Reply:
We haven't converged on single one specifically yet AFAIK, 
```python
def _download_file(
    downloader, path, size=None, mtime=None, existing="error", digests=None
):
```
so `digests` dictionary is provided.  You could just store all of them and use any which was provided (that function ATM estimates them all).  I would not hardcode any specific value in the code.  just perform initial download, get the checksums, use in subsequent "partial re-download" tests.

Reply:
I meant that the test function needs a way to get the checksum so that it can store it in the `.dandidownload` folder when setting up the resumption test.

Reply:
It can just run Digester with desired checksums on that file it (pre)downloaded, or if it is done with internal "fixture" instance -- it could be the file it uploads.

Reply:
Test written & passes.